File: bins/ppos/src/shell.rs
use alloc::boxed::Box;
use bsp::memory::INIT_RAMFS_BASE;
use cpio::CPIOArchive;
use library::{console, format, print, println, string::String};

use crate::driver::{self, mailbox};

pub struct Shell {
    input: String,
}

impl Default for Shell {
    fn default() -> Self {
        Shell::new()
    }
}

impl Shell {
    pub fn new() -> Self {
        Self {
            input: String::from(""),
        }
    }

    fn shell_hint(&self) {
        print!("# ");
    }

    pub fn run(&mut self) -> ! {
        self.shell_hint();
        loop {
            let c = console::console().read_char();
            match c {
                '\r' | '\n' => {
                    self.execute_command();
                    self.shell_hint();
                }
                '\x08' | '\x7f' => self.backspace(),
                ' '..='~' => self.press_key(c),
                _ => (),
            }
        }
    }

    fn help(&self) {
        println!("help\t: print this help menu");
        println!("hello\t: print Hello World!");
        println!("reboot\t: reboot the device");
        println!("cancel-reboot\t: cancel reboot");
        println!("info\t: get hardware infomation");
        println!("ls\t: list files");
        println!("cat\t: show file content");
    }

    fn reboot(&self) {
        driver::watchdog().reset(0x20);
    }

    fn cancel_reboot(&self) {
        driver::watchdog().cancel_reset();
    }

    fn hello(&self) {
        println!("Hello World!");
    }

    fn info(&self) {
        let mem_info = mailbox().get_arm_memory();
        println!("board revision: {:#08x}", mailbox().get_board_revision());
        println!("ARM memory base address: {:#08x}", mem_info.base_address);
        println!("ARM memory size: {} bytes", mem_info.size);
    }

    fn execute_command(&mut self) {
        println!();
        let input = self.input.trim();
        let mut split_result = input.split(" ");
        if let Some(cmd) = split_result.next() {
            let args = split_result.collect::<Box<[&str]>>();
            match cmd {
                "help" => self.help(),
                "hello" => self.hello(),
                "reboot" => self.reboot(),
                "cancel-reboot" => self.cancel_reboot(),
                "info" => self.info(),
                "ls" => self.ls(),
                "cat" => self.cat(args),
                "" => (),
                cmd => println!("{}: command not found", cmd),
            }
        }
        self.input.clear();
    }

    fn press_key(&mut self, key: char) {
        self.input.push(key);
        print!("{}", key);
    }

    fn backspace(&mut self) {
        if self.input.is_empty() {
            return;
        }
        self.input.pop();
        // move the cursor to the previous character and overwrite it with a space
        // then move the cursor back again
        print!("\x08 \x08");
    }

    fn ls(&self) {
        let mut cpio_archive = unsafe { CPIOArchive::from_memory(INIT_RAMFS_BASE) };
        while let Some(file) = cpio_archive.read_next() {
            println!("{}", file.name);
        }
    }

    fn cat(&self, args: Box<[&str]>) {
        if args.len() != 1 {
            println!("Usage: cat <file>");
            return;
        }

        let t = format!("{}\0", args[0]);
        let filename = t.as_str();
        let mut cpio_archive = unsafe { CPIOArchive::from_memory(INIT_RAMFS_BASE) };
        while let Some(file) = cpio_archive.read_next() {
            if file.name == filename {
                for byte in file.content {
                    print!("{}", *byte as char);
                }
                return;
            }
        }
        println!("cat: {}: No such file or directory", filename);
    }
}

File: lab1/shell.c
#include "header/shell.h"
#include "header/uart.h"
#include "header/utils.h"
#include "header/mailbox.h"
#include "header/reboot.h"

void shell(){
  char input_space[256];
  char* input_string = input_space;
  while(1) {
     char element;
     uart_send_string("# ");
     while(1) {
       element = uart_get_char();
       *input_string++ = element;
       if(element == '\n')
       {
        uart_send_string("\r\n");
        *input_string = '\0';
        break;
       }
       else
       {
          uart_send_char(element);
       }
     }
     
     input_string = input_space;
     if(string_compare(input_string,"help")) 
     {
       uart_send_string("help	:print this help menu\n");
       uart_send_string("hello	:print Hello World!\n");
       uart_send_string("info	:Get the hardware's information\n");
       uart_send_string("reboot	:reboot the device\n");
     } 
     else if (string_compare(input_string,"hello")) 
     {
       uart_send_string("Hello World!\n");
     } 
     else if (string_compare(input_string,"info")) 
     {
        get_board_revision();
        get_arm_mem();
     } 
     else if (string_compare(input_string,"reboot")) {
        uart_send_string("Rebooting....\r\n");
        reset(1000);
        break;
     }     
  }
}
